internal/monitoring/watcher_service.go

```diff
@@ -0,0 +1,141 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+// SlackNotifier allows sending notification about messages to Slack channel.
+type SlackNotifier interface {
+	Notify(id, header, details string) error
+}
+
+// WatcherService allows to watch events for a given Pod and send notification to Slack if received event Type is different that `Normal`
+type WatcherService struct {
+	coreCli       corev1.CoreV1Interface
+	slackNotifier SlackNotifier
+	log           logrus.FieldLogger
+
+	watchedObj map[string]*watchObj
+}
+
+type watchObj struct {
+	watcher   watch.Interface
+	sendEvent map[types.UID]struct{}
+}
+
+// NewWatcherService returns new instance of the WatcherService
+func NewWatcherService(eventCli corev1.CoreV1Interface, slackNotifier SlackNotifier, log logrus.FieldLogger) *WatcherService {
+	return &WatcherService{
+		coreCli:       eventCli,
+		slackNotifier: slackNotifier,
+		log:           log.WithField("service", "event:watcher"),
+
+		watchedObj: make(map[string]*watchObj),
+	}
+}
+
+// Register registers given obj and starts watching events from it.
+// Error is not returned when object with the same Name is already registered
+func (s *WatcherService) Register(ref *v1.ObjectReference) error {
+	if _, found := s.watchedObj[ref.Name]; found { // TODO add ns context
+		return nil
+	}
+
+	nsEventCli := s.coreCli.Events(ref.Namespace)
+
+	selector := nsEventCli.GetFieldSelector(s.selectorArgsFromRefObj(ref))
+	eventWatcher, err := nsEventCli.Watch(metav1.ListOptions{
+		FieldSelector: selector.String(),
+	})
+	if err != nil {
+		return errors.Wrapf(err, "while creating watch Event interface for object %q", ref.Name)
+	}
+
+	go s.startWatching(ref, eventWatcher.ResultChan())
+	s.watchedObj[ref.Name] = &watchObj{
+		watcher:   eventWatcher,
+		sendEvent: make(map[types.UID]struct{}),
+	}
+
+	return nil
+}
+
+func (s *WatcherService) startWatching(ref *v1.ObjectReference, events <-chan watch.Event) {
+	// We will also get the restart events here, so we are not checking the status of the pod directly
+	for r := range events {
+		event := r.Object.(*v1.Event)
+		if event.Type != "Normal" {
+			failLogger := s.log.WithField("ID", event.GetUID())
+			obj := s.watchedObj[ref.Name]
+			if _, found := obj.sendEvent[event.GetUID()]; found {
+				continue
+			}
+
+			eventMsg := fmt.Sprintf("Event type: %s, reason: %s, message: %s", event.Type, event.Reason, event.Message)
+			dumpedLogs, err := s.podLogs(ref)
+			if err != nil {
+				failLogger.Errorf("Got error while getting log from pod: %v", err)
+			}
+
+			failureReasonHeader := fmt.Sprintf("*[Phase: MONITORING]* _Discover that Pod %s has problems_", ref.Name)
+			err = s.slackNotifier.Notify(string(event.UID), failureReasonHeader, eventMsg)
+			if err != nil {
+				failLogger.Errorf("Got error while sending notification to Slack: %v", err)
+				continue
+			} else {
+				obj.sendEvent[event.UID] = struct{}{}
```

Comment:
the GC is doing that. 
I've made such assumption:
I have a map where we putting the watched Pods
`watchedObj map[string]*watchObj`

the key is the Pod name, and the value is the struct with `sendEvent` map and with the `watch.Interface`

in `sendEvent` map I'm saving the id of the given UID, I've used as a value the empty struct which in Go empty struct has a width of zero ( It occupies zero bytes of storage). And this maps should not have many entries because do you expect that more than 1000 different events will be sent to Pod?

and whole that things are released when we are stopping watching given Pod.